compiler/src/dotty/tools/dotc/typer/Synthesizer.scala

```diff
@@ -379,7 +379,7 @@ class Synthesizer(typer: Typer)(using @constructorOnly c: Context):
                 // avoid type aliases for tuples
                 Right(MirrorSource.GenericTuple(types))
               case _ => reduce(tp.underlying)
-          case tp: MatchType => reduce(tp.normalized)
+          case tp: MatchType => reduce(tp.tryNormalize.orElse(tp.superType))
```

Comment:
This is actually not quite correct, if I change the definition of DoesNotReduce to:
```
type DoesNotReduce[T] = T match
  case String => T
```
Then it infers an upper-bound of `T` and `tp.superType` ends up being `Option[Int]` and so a mirror is found when none should be found. This seems like a more general problem with `superType` though, so maybe something we should address after fixing this regression.

Reply:
Or perhaps we should never summon the mirror if the match type cannot be reduced? Maybe other situations where we use `superType` are actually fine. /cc @sjrd

Reply:
I agree using `superType` looks a bit suspicious given that `Mirror.Of[_]` isn't covariant.
But then again, it also used right below.